### ISSUE TITLE
tracking reputation

### ISSUE THREAD

**Issue by adrivas** [HarpyWar/nfsuserver#6]
Hi, I don't know if you still care about this project or you've abandoned it, but I was wondering if reputation tracking is possible.
I've looked over the code to change it, but I still don't understand it fully. I'm not asking for an implementation, but could you give me some advice?

**Comment by HarpyWar** [comment]
I did not write the server code, just upgraded it a little. But as I can see all income server packets are handled here:
https://github.com/HarpyWar/nfsuserver/blob/master/nfsuserver/nfsuserver2.cpp#L454

May be game stats reported by each player on game delete event
https://github.com/HarpyWar/nfsuserver/blob/master/nfsuserver/nfsuserver2.cpp#L504
Or in this rank packet
https://github.com/HarpyWar/nfsuserver/blob/master/nfsuserver/nfsuserver2.cpp#L796

Probably there are not all packets implemented and you may need to debug that part of code with break points. 
Or try to enable logging feature and analyse whats going on in `traffic.log` when a game ends.
```
# enable debug logging [0|1]
verbose = 1

# enable server traffic logging in a separate file
logalltraffic = 1
```

**Comment by adrivas** [comment]
Thanks
I've been debugging that listener and for some reason I did not check the game based events/commands (the ones that start with "g"). Unfortunately I still haven't figured out the parameter that represents the reputation (I've tried a few, even ones that were already in the code).
I'm taking a break from this, but I will try again some other time. If I figure something out, I'll put it here.

**Comment by ian-travers** [comment]
@adrivas, you have to process a RANK replica correctly. 

Every player sends RANK replica to the server when a race finished. It looks like:
`rank      _ЮAUTH=
WHEN=2003.12.8 15:52:54
REPT=lenusik
NAME0=lenusik
NAME1=stopcheater
RESU=AgMABRwACgEAAQm7RRUAQYj2RgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPgUkgT42uj9APC1HAAAAAAAAAAC/gAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQCABAqQgAAAgo+qq6PQaFZmgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQYh89UHSPIZDrCyjAAAAAAAAAAC/gAAAAAABYAAAAWAAAAFgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAhEicg%3d%3d` 

REPT – author (sender)
NAME0 = lenusik – player, seeded No. 0 (important)
NAME1 = stopcheater – player, seeded No. 1
RESU – base64 encoded string with result info.

Here is what I was able to decipher.

Decoded RESU length =  7 + P(58 + 8 * L), where L – laps, P – players. 
(L = 1 for sprint always, L = 2 for drag always, L = 2..10 for circuit and drift)

First 7 bytes – is a header with common race info. 
| offset | length | description | possible values |
|----|----|----|----|
| +00 | 1 | Players count |  02 .. 04 |
| +01 | 1 | Race type | 0 - circuit, 1 - sprint, 2 - drag, 3 - drift |
| +02 | 1 | REPT seeded No | 00 .. 03 |
| +03 | 2 | Track ID |  see below |
| +05 | 1 | Direction | 0 -forward, 1 - reverse |
| +06 | 1 | Laps | 02 .. 0A - circuit and drift, 01 - sprint, 02- drag |

**block for each player**:
+00 (int 1)  – seed No.  NAMEx=persona_name. Now we can get persona (or username in the game interface)!
+01 (int 1) – ???
+02 (int 1) – place. 01 = first place; 02 = second place …
+03 (int 1) –  finish flag. 01– finished; 06 – engine blown (drag only) not finished; 09..0B (probably) – not finished
+04 (float 4) – distance in laps. If player finished then ≈ L 
+08 (float 4) – full time in sec. 
+12 (float 4) – best lap in sec.
+12 + (i * 4) (float 4) – lap #i time in sec. (i = 1..L)
+12 + (i * 4) + 4 (float 4) – ???
+12 + (i * 4) + 8 (float 4) – ???
+12 + (i * 4) + 12 (float 4) – distance in meters
+12 + (i * 4) + 16 (float 4) – ???
+12 + (i * 4) + 20 (float 4) – ???
+12 + (i * 4) + 24 (float 4) – disconnect flag (probably) if = –1 (attention! float value). 
+12 + (i * 4) + 28 (int 4) – total drift score
+12 + (i * 4) + 32 (int 4) – best continuous drift score
+12 + (i * 4) + 32 + (i * 4) (int 4) – drift score on lap #i (i = 1..L)
+12 + (i * 4) + 32 + (i * 4) + 10 (int 1) – car ID
 
As you can see, I was unable to decipher several parameters

**Comment by ian-travers** [comment]
Tracks:
| hex | dec | name | hex | dec | name |
|---|---|---|---|---|---|
| 03E9 | 1001 | Market Street | 0455 | 1109 | 7th Sparling |
| 03EA | 1002 | Stadium | 04B1 | 1201 | 14th and Vine Construction |
| 03EB | 1003 | Olympic Square | 04B1 | 1202 | Highway 1 |
| 03EC | 1004 | Terminal | 04B6 | 1206 | Main Street |
| 03ED | 1005 | Atlantica | 04B7 | 1207 | Commercial |
| 03EE | 1006 | Inner City | 04BA | 1210 | 14th and Vine |
| 03EF | 1007 | Port Royal | 04BE | 1214 | Main Street Construction |
| 03F0 | 1008 | National Rail | 0515 | 1301 | Drift Track 1 |
| 044E | 1102 | Liberty Gargens | 0516 | 1302 | Drift Track 2 |
| 044F | 1103 | Broadway | 0517 | 1303 | Drift Track 3 |
| 0450 | 1104 | Lock up | 0518 | 1304 | Drift Track 4 |
| 0451 | 1105 | 9th Frey | 0519 | 1305 | Drift Track 5 |
| 0452 | 1106 | Bedard Bridge | 051A | 1306 | Drift Track 6 |
| 0453 | 1107 | Spillway | 051B | 1307 | Drift Track 7 |
| 0454 | 1108 | 1st Ave. Truck Stop | 051C | 1308 | Drift Track 8 |

Cars:
| ID | code | name | ID | code | name |
|---|---|---|---|---|---|
| 00 | A | VW Golf GTI | 0A | K | Honda S2000 |
| 01 | B | Ford Focus | 0B | L | Acura RSX |
| 02 | C | Mazda Miata MX-5 | 0C | M | Subaru Impreza |
| 03 | D | Dodge Neon | 0D | N | Mitsibisji Lancer |
| 04 | E | Honda Civic | 0E | O | Acura Integra TypeR |
| 05 | F | Peugeot 206 | 0F | P | Hyundai Tiburon GT |
| 06 | G | Toyota Celica | 10 | Q | Nissan 350Z |
| 07 | H | Mitsubishi Eclipse | 11 | R | Nissan Sentra SER | 
| 08 | I | Mazda RX-7 | 12 | S | Nissan 240SX |
| 09 | J | Toyota Supra | 13 | T | Nissan Skyline GTR |

**Comment by DanielCarter1615** [comment]
Sorry for hijacking this thread. Will there be a new release soon with all the recent changes and fixes?

**Comment by DanielCarter1615** [comment]
@ian-travers I tested the new server code, but somehow the reputation is still not working for me. Do you know why? We drove an ranked race, but the rep didn't change. Also the win - loss stats didn't get updated.

The ranking and week's best performers seems to be working fine though.

**Comment by ian-travers** [comment]
@DanielCarter1615 if you meant the display of reputation in the lobby, then perhaps you may not have the latest changes. Check if the [Fix player REP in the rooms pull request](https://github.com/HarpyWar/nfsuserver/pull/8) is included in your build.

**Comment by DanielCarter1615** [comment]
@ian-travers Yeah, I was using the last commit, but still had issues (not only with the rep in the rooms ,but also with the calculation of it). I will play around a little bit more and try to come up with a detailed bug report

**Comment by HarpyWar** [comment]
> Sorry for hijacking this thread. Will there be a new release soon with all the recent changes and fixes?

Release binaries are published. Please, let us know if everything work or not. 
Note about `rusers.dat` which must be re-created if kept from the previos server version.

**Comment by DanielCarter1615** [comment]
@ian-travers @HarpyWar I think I found two issues:
1. if you join a lobby your rep points will automatically be shown as the rep points for everyone in the lobby. E.g. player 1 has 200 rep, player 2 has 100 rep. If player 1 joins a lobby where player 2 is already in, then for player 1 it looks like player 2 also has 200 rep. But if player 2 joins after player 1 into the lobby, then the rep of player 2 is shown correctly as 100.
2. The other issue: In a 1on1 if a player gets disconnected, because he didn't move, he actually gets granted a win. The person who was normally driving, before the disconnection of the other player, will then get a loss added in the statistics. It should be the other way around.

**Comment by ian-travers** [comment]
It seems the `move` replica is not being handled correctly. I will make a brand new repository based on my old code and add features from @HarpyWar.

**Comment by ian-travers** [comment]
@DanielCarter1615 I've created [nfsu-server](https://github.com/ian-travers/nfsu-server) repository. I hope the rep points issue has been resolved.

**Comment by DanielCarter1615** [comment]
Hey @ian-travers,
thanks, I will check it out tonight!
